Mysterious error with optional dependencies in pyproject.toml

I have been following the guide in the README file on how to use pip-tools with pyproject.toml. Unfortunately, if I add a ``project.optional-dependencies`` section to pyproject.toml, *regardless of what its contents are*, I get the following error:

```
(base) PS C:\Users\adewar\code\Solidity-GUI> pip-compile -v pyproject.toml
C:\ProgramData\Miniconda3\lib\site-packages\_distutils_hack\__init__.py:33: UserWarning: Setuptools is replacing distutils.
  warnings.warn("Setuptools is replacing distutils.")
Creating virtualenv isolated environment...
find interpreter for spec PythonSpec(path=C:\ProgramData\Miniconda3\python.exe)
proposed PythonInfo(spec=CPython3.9.13.final.0-64, exe=C:\ProgramData\Miniconda3\python.exe, platform=win32, version='3.9.13 (main, Oct 13 2022, 21:23:06) [MSC v.1916 64 bit (AMD64)]', encoding_fs_io=utf-8-utf-8)
create virtual environment via CPython3Windows(dest=C:\Users\adewar\AppData\Local\Temp\build-env-m1awcc93, clear=False, no_vcs_ignore=False, global=False)
add seed packages via FromAppData(download=False, pip=bundle, via=copy, app_data_dir=C:\Users\adewar\AppData\Local\pypa\virtualenv)
Installing packages in isolated environment... (setuptools >= 40.8.0, wheel)
Getting build dependencies for wheel...
Backend subprocess exited when trying to invoke get_requires_for_build_wheel
Failed to parse C:\Users\adewar\code\Solidity-GUI\pyproject.toml
```

#### Environment Versions

1. Windows 10 (replicated on Arch Linux)
1. Python version: `python 3.9.13`
1. pip version: `pip 22.2.2 from C:\ProgramData\Miniconda3\lib\site-packages\pip (python 3.9)`
1. pip-tools version: `pip-compile, version 6.9.0`

#### Steps to replicate

1. Add a `project.optional-dependencies` section to your `pyproject.toml`
2. Run `pip-compile pyproject.toml`

A colleague has tracked down the source of the problem. If you check ``pyproject.toml`` with [validate-pyproject](https://pypi.org/project/validate-pyproject) you get:
```
[ERROR] project.authors[{data__authors_x}] must be object
```

Putting the ``authors`` field into the correct format fixes things, i.e.:
```py
authors = [{ name = "author1", email = "author1@email.com" }]
```

--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -32,7 +32,8 @@ repos:
           - pep517==0.10.0
           - toml==0.10.2
           - pip==20.3.4
-          - build==0.9.0
+          - build==1.0.0
+          - pyproject_hooks==1.0.0
   - repo: https://github.com/PyCQA/bandit
     rev: 1.7.5
     hooks:

--- a/piptools/scripts/compile.py
+++ b/piptools/scripts/compile.py
@@ -15,7 +15,7 @@ from click.utils import LazyFile, safecall
 from pip._internal.req import InstallRequirement
 from pip._internal.req.constructors import install_req_from_line
 from pip._internal.utils.misc import redact_auth_from_url
-from pyproject_hooks import default_subprocess_runner
+from pyproject_hooks import default_subprocess_runner, quiet_subprocess_runner
 
 from .._compat import parse_requirements
 from ..cache import DependencyCache
@@ -320,11 +320,14 @@ def cli(
             constraints.extend(reqs)
         elif is_setup_file:
             setup_file_found = True
+            build_runner = quiet_subprocess_runner
+            if verbose:
+                build_runner = default_subprocess_runner
             try:
                 metadata = project_wheel_metadata(
                     os.path.dirname(os.path.abspath(src_file)),
                     isolated=build_isolation,
-                    runner=default_subprocess_runner,
+                    runner=build_runner,
                 )
             except BuildBackendException as e:
                 log.error(str(e))

--- a/piptools/scripts/compile.py
+++ b/piptools/scripts/compile.py
@@ -320,14 +320,15 @@ def cli(
             constraints.extend(reqs)
         elif is_setup_file:
             setup_file_found = True
-            build_runner = quiet_subprocess_runner
-            if verbose:
-                build_runner = default_subprocess_runner
             try:
                 metadata = project_wheel_metadata(
                     os.path.dirname(os.path.abspath(src_file)),
                     isolated=build_isolation,
-                    runner=build_runner,
+                    runner=(
+                        default_subprocess_runner
+                        if verbose
+                        else quiet_subprocess_runner
+                    ),
                 )
             except BuildBackendException as e:
                 log.error(str(e))

--- a/piptools/scripts/compile.py
+++ b/piptools/scripts/compile.py
@@ -11,6 +11,7 @@ from typing import IO, Any, BinaryIO, cast
 import click
 from build import BuildBackendException
 from build.util import project_wheel_metadata
+from pyproject_hooks import default_subprocess_runner
 from click.utils import LazyFile, safecall
 from pip._internal.req import InstallRequirement
 from pip._internal.req.constructors import install_req_from_line
@@ -323,6 +324,7 @@ def cli(
                 metadata = project_wheel_metadata(
                     os.path.dirname(os.path.abspath(src_file)),
                     isolated=build_isolation,
+                    runner=default_subprocess_runner,
                 )
             except BuildBackendException as e:
                 log.error(str(e))

--- a/piptools/scripts/compile.py
+++ b/piptools/scripts/compile.py
@@ -11,11 +11,11 @@ from typing import IO, Any, BinaryIO, cast
 import click
 from build import BuildBackendException
 from build.util import project_wheel_metadata
-from pyproject_hooks import default_subprocess_runner
 from click.utils import LazyFile, safecall
 from pip._internal.req import InstallRequirement
 from pip._internal.req.constructors import install_req_from_line
 from pip._internal.utils.misc import redact_auth_from_url
+from pyproject_hooks import default_subprocess_runner
 
 from .._compat import parse_requirements
 from ..cache import DependencyCache

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -36,7 +36,8 @@ classifiers = [
 keywords = ["pip", "requirements", "packaging"]
 dependencies = [
   # direct dependencies
-  "build",
+  "build >= 1.0.0",
+  "pyproject_hooks",
   "click >= 8",
   "pip >= 22.2",
   "tomli; python_version < '3.11'",

--- a/tests/test_cli_compile.py
+++ b/tests/test_cli_compile.py
@@ -2603,7 +2603,8 @@ def test_error_in_pyproject_toml(fake_dists, runner, make_module, capfd):
         cli, ["-n", "--no-build-isolation", "-v", "--find-links", fake_dists, meta_path]
     )
     assert out.exit_code == 2, out.stderr
-    assert "`project` must contain ['name'] properties" in capfd.readouterr().err
+    captured = capfd.readouterr()
+    assert "`project` must contain ['name'] properties" in captured.err
 
 
 @pytest.mark.network

--- a/tests/test_cli_compile.py
+++ b/tests/test_cli_compile.py
@@ -2582,7 +2582,10 @@ def test_input_formats(fake_dists, runner, make_module, fname, content):
     assert "extra ==" not in out.stderr
 
 
-def test_error_in_pyproject_toml(fake_dists, runner, make_module, capfd):
+@pytest.mark.parametrize("verbose_option", (True, False))
+def test_error_in_pyproject_toml(
+    fake_dists, runner, make_module, capfd, verbose_option
+):
     """
     Test that an error in pyproject.toml is reported.
     """
@@ -2594,17 +2597,23 @@ def test_error_in_pyproject_toml(fake_dists, runner, make_module, capfd):
         """
     )
     meta_path = make_module(fname=fname, content=invalid_content)
-    out = runner.invoke(
-        cli, ["-n", "--no-build-isolation", "--find-links", fake_dists, meta_path]
+
+    options = []
+    if verbose_option:
+        options = ["--verbose"]
+
+    options.extend(
+        ["-n", "--no-build-isolation", "--find-links", fake_dists, meta_path]
     )
-    assert out.exit_code == 2, out.stderr
-    assert "`project` must contain ['name'] properties" not in capfd.readouterr().err
-    out = runner.invoke(
-        cli, ["-n", "--no-build-isolation", "-v", "--find-links", fake_dists, meta_path]
-    )
+
+    out = runner.invoke(cli, options)
+
     assert out.exit_code == 2, out.stderr
     captured = capfd.readouterr()
-    assert "`project` must contain ['name'] properties" in captured.err
+
+    assert (
+        "`project` must contain ['name'] properties" in captured.err
+    ) is verbose_option
 
 
 @pytest.mark.network

--- a/tests/test_cli_compile.py
+++ b/tests/test_cli_compile.py
@@ -2589,9 +2589,9 @@ def test_error_in_pyproject_toml(fake_dists, runner, make_module, capfd):
     fname = "pyproject.toml"
     invalid_content = dedent(
         """\
-            [project]
-            invalid = "metadata"
-            """
+        [project]
+        invalid = "metadata"
+        """
     )
     meta_path = make_module(fname=fname, content=invalid_content)
     out = runner.invoke(

--- a/tests/test_cli_compile.py
+++ b/tests/test_cli_compile.py
@@ -2583,6 +2583,37 @@ def test_input_formats(fake_dists, runner, make_module, fname, content):
 
 
 @pytest.mark.network
+def test_error_in_pyproject_toml(fake_dists, runner, make_module, capfd):
+    """
+    Test that an error in pyproject.toml is reported.
+    """
+    fname = "pyproject.toml"
+    content = """
+            [build-system]
+            requires = ["flit_core >=2,<4"]
+            build-backend = "flit_core.buildapi"
+
+            [tool.flit.metadata]
+            module = "sample_lib"
+            author = "Vincent Driessen"
+            author-email = "me@nvie.com"
+            requires = ["small-fake-a==0.1"]
+    """
+    content += """
+            maintainer = ["Vincent Driessen <nvie>"]  # error is here
+    """
+    meta_path = make_module(fname=fname, content=content)
+    out = runner.invoke(
+        cli, ["-n", "--no-build-isolation", "--find-links", fake_dists, meta_path]
+    )
+    assert out.exit_code == 2, out.stderr
+    assert (
+        "ConfigError: Expected a string for maintainer field, found"
+        in capfd.readouterr().err
+    )
+
+
+@pytest.mark.network
 @pytest.mark.parametrize(("fname", "content"), METADATA_TEST_CASES)
 def test_one_extra(fake_dists, runner, make_module, fname, content):
     """

--- a/tests/test_cli_compile.py
+++ b/tests/test_cli_compile.py
@@ -2582,35 +2582,28 @@ def test_input_formats(fake_dists, runner, make_module, fname, content):
     assert "extra ==" not in out.stderr
 
 
-@pytest.mark.network
 def test_error_in_pyproject_toml(fake_dists, runner, make_module, capfd):
     """
     Test that an error in pyproject.toml is reported.
     """
     fname = "pyproject.toml"
-    content = """
-            [build-system]
-            requires = ["flit_core >=2,<4"]
-            build-backend = "flit_core.buildapi"
-
-            [tool.flit.metadata]
-            module = "sample_lib"
-            author = "Vincent Driessen"
-            author-email = "me@nvie.com"
-            requires = ["small-fake-a==0.1"]
-    """
-    content += """
-            maintainer = ["Vincent Driessen <nvie>"]  # error is here
-    """
-    meta_path = make_module(fname=fname, content=content)
+    invalid_content = dedent(
+        """\
+            [project]
+            invalid = "metadata"
+            """
+    )
+    meta_path = make_module(fname=fname, content=invalid_content)
     out = runner.invoke(
         cli, ["-n", "--no-build-isolation", "--find-links", fake_dists, meta_path]
     )
     assert out.exit_code == 2, out.stderr
-    assert (
-        "ConfigError: Expected a string for maintainer field, found"
-        in capfd.readouterr().err
+    assert "`project` must contain ['name'] properties" not in capfd.readouterr().err
+    out = runner.invoke(
+        cli, ["-n", "--no-build-isolation", "-v", "--find-links", fake_dists, meta_path]
     )
+    assert out.exit_code == 2, out.stderr
+    assert "`project` must contain ['name'] properties" in capfd.readouterr().err
 
 
 @pytest.mark.network

